docs/redesign-specs/REST/Api.md

```diff
@@ -0,0 +1,1522 @@
+  * Administrative access
+  * No special access
+ In general, administrative access is sufficient to perform any action.
+* A list of possible errors.
+* A user should only perform an `<action>` on a `<resource>` that he owns or is shared with him by another user.
+* `Admin` is a privileged user who can perform an `<action>` on any `<resource>` that belongs to any user.
+
+#### Possible error codes for BMI API:
+
+* 400 - Invalid Parameters
+* 401 - Invalid Credentials
+* 403 - Insufficient Permissions
+* 404 - Resource Does Not Exist
+* 409 - Resource Busy
+* 423 - Duplicate Operation
+* 500 - Server/Internal Error
+
+Below are examples how to document a BMI REST API call:
+
+`<HTTP-method-type> /<resource>/<Id>`
+
+**OR**
+
+`<HTTP-method-type> /<resource>/<Id>/<action>`
+
+The goal is to have unique `REST method URL`. Therefore, for some cases we should pass `<action>` as part of URL too.
+
+**NOTE**: `HTTP-method-type` can be `POST`, `PUT`, `PATCH`, `GET` or `DELETE`.
+* `POST` is used when a new instance of a resource is created, therefore there is no `Id` in the API call. The response body should include the `Id` for the new instance.
+
+* `PUT` is used when an existent instance needs to be changed/updated (all attributes will be updated), it returns a status indicating if it was successful or not.
+
+* `PATCH` is used when an existent instance needs to be changed/updated slightly (a few attributes will be updated), it returns a status indicating if it was successful or not.
+
+* `GET` is used to read an instance or list of instances. The response will be in JSON format in case of success and otherwise an error message.
+
+* `DELETE` is used for deleting an instance, and its response will a status indicating if it was successful or not.
+
+#### Request Body Example (Optional - Depending on `<HTTP-method-type>`):
+
+    {
+                "idOfTheNewInstance": "a name for new instance",
+                "someField": "a value",
+                "thisIsAnExample": true,
+                "someOptionalField (Optional): {
+                        "more-fields": 12352356,
+                        ...
+                }
+    }
+
+**NOTE**: After the example, an explanation for each parameter must be provided.
+
+The response header will contain the HTTP status code:
+* 200 for success
+* 400, 404, etc for failure
+
+#### Response Body Example (Success - Optional):
+
+    {
+        "someInfo": "Hello, World!",
+        "numbers": [1,2,3]
+    }
+
+**Response Parameters**:
+* `someInfo`: Expected `response` explanation.
+* `numbers` : More explanation.
+
+#### Response Body Example (Failure):
+
+    {
+        "errorMessage": "This is the reason for the encountered error"
+    }
+
+**NOTE**: `Response Body Example (Failure)` is not required for every API call
+
+#### Authentication (Optional):
+* If Authentication is required by REST API call then the Credentials/Token should be passed using `HTTP Basic Auth`.
```

Comment:
Different authentication methods have different ways to pass credentials. Please, please, do not do this. Use a middleware (like keystonemiddleware) or apache module for this.